odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1516 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+from numbers import Number
+
+from odl.operator.operator import (
+    Operator, OperatorComp, OperatorLeftScalarMult, OperatorRightScalarMult,
+    OperatorRightVectorMult, OperatorSum)
+from odl.operator.default_ops import (ResidualOperator, IdentityOperator,
+                                      ConstantOperator, ScalingOperator)
+from odl.set.space import LinearSpaceVector
+from odl.solvers.advanced import (proximal_arg_scaling, proximal_translation,
+                                  proximal_quadratic_perturbation)
+
+
+# TODO: Add missing functionals here
+__all__ = ('Functional', 'ConvexConjugateArgScaling',
+           'FunctionalLeftScalarMult',
+           'ConvexConjugateFuncScaling', 'ConvexConjugateLinearPerturb',
+           'ConvexConjugateTranslation', 'TranslatedFunctional')
+
+
+class Functional(Operator):
+    """Implementation of a functional class.
+
+    Notes
+    -----
+    Note that the implementation of the functional class has assued that the
+    functionals are defined on a Hilbert space, i.e., that we consider
+    functionals
+
+    .. math::
+
+            f : X \\to F,
+
+    where :math:`X` is a Hilbert space and :math:`F` is a field of scalars
+    associated with :math:`X`. This has been done in order to simplify the
+    concept of *convex conjugate functional*. Since Hilbert spaces as selfdual
+    the convex conjugate functional is defined as
+
+    .. math::
+
+            f^* : X \\to F,
+
+            f^*(x^*) = \\sup_{x} \{ \\langle x^*,x \\rangle - f(x)  \}.
```

Comment:
This is in fact the definition for the case that `F` is the set of real numbers. Perhaps we should mention that all definitions are with real numbers in mind, but that other fields are not forbidden, just no guarantee that things are correct any longer.